src/closures/query_channel.rs

```diff
@@ -0,0 +1,227 @@
+        send: z_owned_closure_owned_query_t::empty(),
+        recv: z_owned_query_channel_closure_t::empty(),
+    }
+}
+
+/// Creates a new blocking fifo channel, returned as a pair of closures.
+///
+/// If `bound` is different from 0, that channel will be bound and apply back-pressure when full.
+///
+/// The `send` end should be passed as callback to a `z_get` call.
+///
+/// The `recv` end is a synchronous closure that will block until either a `z_owned_query_t` is available,
+/// which it will then return; or until the `send` closure is dropped and all queries have been consumed,
+/// at which point it will return an invalidated `z_owned_query_t`, and so will further calls.
+#[no_mangle]
+pub extern "C" fn zc_query_fifo_new(bound: usize) -> z_owned_query_channel_t {
+    if bound == 0 {
+        let (tx, rx) = std::sync::mpsc::channel();
+        z_owned_query_channel_t {
+            send: From::from(move |query: &mut z_owned_query_t| {
+                if let Some(query) = query.take() {
+                    if let Err(e) = tx.send(query) {
+                        log::error!("Attempted to push onto a closed query_fifo: {}", e)
+                    }
+                }
+            }),
+            recv: From::from(move |receptacle: &mut z_owned_query_t| {
+                *receptacle = match rx.recv() {
+                    Ok(val) => val.into(),
+                    Err(_) => None.into(),
+                };
+                true
+            }),
+        }
+    } else {
+        let (tx, rx) = std::sync::mpsc::sync_channel(bound);
+        z_owned_query_channel_t {
+            send: From::from(move |query: &mut z_owned_query_t| {
+                if let Some(query) = query.take() {
+                    if let Err(e) = tx.send(query) {
+                        log::error!("Attempted to push onto a closed query_fifo: {}", e)
+                    }
+                }
+            }),
+            recv: From::from(move |receptacle: &mut z_owned_query_t| {
+                *receptacle = match rx.recv() {
+                    Ok(val) => val.into(),
+                    Err(_) => None.into(),
+                };
+                true
+            }),
+        }
+    }
+}
+
+/// Creates a new non-blocking fifo channel, returned as a pair of closures.
+///
+/// If `bound` is different from 0, that channel will be bound and apply back-pressure when full.
+///
+/// The `send` end should be passed as callback to a `z_get` call.
+///
+/// The `recv` end is a synchronous closure that will block until either a `z_owned_query_t` is available,
+/// which it will then return; or until the `send` closure is dropped and all queries have been consumed,
+/// at which point it will return an invalidated `z_owned_query_t`, and so will further calls.
+#[no_mangle]
+pub extern "C" fn zc_query_non_blocking_fifo_new(bound: usize) -> z_owned_query_channel_t {
+    if bound == 0 {
+        let (tx, rx) = std::sync::mpsc::channel();
```

Comment:
It seems a lot of code replication just for `std::sync::mpsc::channel()` vs `std::sync::mpsc::sync_channel(bound)` channel creation.